aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -859,6 +860,99 @@ std::pair<Tensor, Tensor> _not_zero_mask_to_col_row_indices(
 
 // Sparse layout conversions Start
 
+static inline
+void _to_sparse_check_arguments(const std::string& funcname, const Tensor& self, const int64_t sparse_dim) {
+  auto layout_from = self.layout();
+  auto layout_to = kSparse;
+
+  if (layout_from == kStrided) {
+    if (sparse_dim < 0 || (sparse_dim == 0 && self.dim() > 0)) {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " with sparse_dim argument <0, or ==0 when self.dim()>0 is not supported");
+    }
+    if (sparse_dim > self.dim()) {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " with sparse_dim argument >self.dim() is not supported");
+    }
+  } else if (layout_from == kSparse) {
+    if (sparse_dim != self.sparse_dim()) {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " with sparse_dim argument !=self.sparse_dim() is not supported");
+    }
+  } else {
+    if (sparse_dim != 2) {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " with sparse_dim argument !=2 is not supported");
+    }
+  }
+}
+
+static inline
+void _to_sparse_check_arguments(const std::string& funcname, const Tensor& self, c10::optional<c10::Layout> layout, OptionalIntArrayRef blocksize, c10::optional<int64_t> dense_dim_opt) {
+  auto layout_from = self.layout();
+  auto layout_to = layout.value_or(kSparse);
+
+  if (layout_to != kSparse && layout_to != kSparseCsr && layout_to != kSparseCsc && layout_to != kSparseBsr && layout_to != kSparseBsc) {
+    AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " is not supported");
+  }
+
+  if (layout_from == kSparse && layout_to != kSparse) {
+    if (self.sparse_dim() != 2) {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " for input tensors with sparse_dim()!=2 is not supported");
+    }
+  }
+
+  if ((layout_from == kSparseCsr && layout_to == kSparseBsr)
+      || (layout_from == kSparseCsc && layout_to == kSparseBsc)) {
+    if (self.dim() != 2 + self.dense_dim()) {
+      AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " for batched inputs is not supported");
+    }
+  }
+
+  if (blocksize.has_value()) {
+    auto blocksize_to = *blocksize;
+    if (blocksize_to[0] <= 0 || blocksize_to[1] <= 0) {
+      AT_ERROR(funcname, ": blocksize needs to be positive, but got ", blocksize_to);
+    }
+
+    if (layout_to == kSparseBsr || layout_to == kSparseBsc) {
+      if (layout_from == kSparseBsr || layout_from == kSparseBsc) {
+        auto blocksize_from = at::sparse_csr::getBlockSize(self);
+        if (!(blocksize_to == blocksize_from)) {
+          AT_ERROR(funcname, ": conversion from ", layout_from, " to ", layout_to, " with blocksize changed from ", blocksize_from, " to ", blocksize_to, " is not supported");
+        }
+      } else {
+        auto dense_dim = (layout_from == kStrided) ? dense_dim_opt.value_or(0) : self.dense_dim();
```

Comment:
Not for now, namely `dense_dim_opt` works only when input tensor layout is `kStrided`; somewhat below these lines, there is a check that would report an error if `dense_dim_opt` has value and input layout is not `kStrided`.

In general, it would be the best to have conversion routines that make it possible to change everything at will: number of batch dims, number of dense dims, block sizes etc.  But we already have number of limitations at the moment, like this one with `dense_dim` argument being accepted only for dense input tensors.  I'm trying with this PR to enumerate these limitations at the single place, at least for conversions to sparse formats.

Reply:
Ok, makes sense. However, silently ignoring the user-specified `dense_dim` argument for non-strided inputs is still a bug. Could you add a check for this?

Reply:
It won't be ignored, an error will be reported.  It's just that the checks for `dense_dim_opt` argument are slightly below in the code, after all the checks related to the `blocksize` argument.

Here is an example, showing that an error gets reported (ignore that this particular `dense_dim` value don't have sense anyway):
```python
coo = torch.eye(2).to_sparse()
print(coo.to_sparse_csr(dense_dim=1))
```